### PR TITLE
statistics: add report list

### PR DESCRIPTION
**Pull request by jma**
* Adds updated and created date.
* Adds filter by libraries.
* Adds live values.
* Adds report list.

### CODE REVIEW

File: projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/reports-list/reports-list.component.html [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  GNU Affero General Public License for more details.
+
+  You should have received a copy of the GNU Affero General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+-->
+
+<table *ngIf="data?.length; else noResult" class="mt-4 table table-striped">
+  <tbody>
+      <tr *ngFor="let v of data">
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/reports-list/reports-list.component.html:20]: Set a more explicit variable name

File: projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/report-data/report-data.component.html [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  GNU Affero General Public License for more details.
+
+  You should have received a copy of the GNU Affero General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+-->
+
+<p-table *ngIf="data && data.length > 0; else noResult" [value]="data" [tableStyle]="{ 'min-width': '50rem' }">
+  {{data}}
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/report-data/report-data.component.html:19]: Missing space `{{ data }}`

File: projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/reports-list/reports-list.component.html [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        <th>
+            {{v.created | dateTranslate: 'longDate' }}
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/reports-list/reports-list.component.html:22]: Missing space

File: projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/reports-list/reports-list.component.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+export class ReportsListComponent implements OnInit {
+  // persistent identifier of the current stat report configuration
+  @Input() pid: any;
+
+  // list of the corresponding reports from elasticsearch
+  data: Array<any>;
+
+  /**
+   * Constructor
+   * @param _recordService - RecordService
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/reports-list/reports-list.component.ts:34]: Missing docstring

File: projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/reports-list/reports-list.component.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    private _apiService: ApiService
+  ) {}
+
+  /**
+   * Get the report item URL
+   *
+   * @param pid - Persistent Identifier value.
+   * @returns the URL as string.
+   */
+  getReportUrl(pid: string) {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/reports-list/reports-list.component.ts:47]: Mssing return type `getReportUrl(pid: string): string {`

File: projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.html [hunk 6]
```diff
@@ -64,8 +72,31 @@ <h5 class="card-header" translate>Configuration</h5>
             <dt class="col-3 label-title" translate>Filter</dt>
             <dd class="col-9">{{ record.metadata.category.indicator.filter }}</dd>
           </ng-container>
+          <ng-container *ngIf="record.created">
+            <dt class="col-3 label-title" translate>Created at</dt>
+            <dd class="col-9">{{ record.created | dateTranslate: 'medium'}}</dd>
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.html:77]: Missing space

File: projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.html [hunk 7]
```diff
@@ -64,8 +72,31 @@ <h5 class="card-header" translate>Configuration</h5>
...
+  <tabset>
+    <tab>
+      <ng-template tabHeading>
+       <span translate>Reports</span>
+      </ng-template>
+        <admin-reports-list [pid]="record.metadata.pid"></admin-reports-list>
+    </tab>
+    <tab (selectTab)="getLiveValues()">
+      <ng-template tabHeading>
+       {{'Live Values' | translate}}
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.html:96]: Missing space

File: projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.ts [hunk 8]
```diff
@@ -14,19 +14,68 @@
...
+
+  /** onDestroy hook */
+  ngOnDestroy(): void {
+    this._subscriptions.unsubscribe();
+  }
+
+  /**
+   * Preview values corresponding to the current configuration.
+   *
+   */
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.ts:69]: `/** Preview values corresponding to the current configuration. */`

File: projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.ts [hunk 9]
```diff
@@ -14,19 +14,68 @@
...
+  getLiveValues() {
+    // only once
+    if (this.liveData != null){
+      return
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.ts:73]: Missing `;`

File: projects/admin/src/app/routes/route-tool.service.ts [hunk 10]
```diff
@@ -238,30 +245,44 @@ export class RouteToolService {
    *      - canUpdate permission
    *      - canDelete permission
    */
-  permissions(record: any, recordType: string, membership = false): Observable<any> {
+  permissions(
+    record: any,
+    recordType: string,
+    membership = false
+  ): Observable<any> {
```
Comment (Garfield-fr) [inline projects/admin/src/app/routes/route-tool.service.ts:252]: 1 line

File: projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.ts [hunk 11]
```diff
@@ -14,19 +14,68 @@
...
+    this._subscriptions.unsubscribe();
+  }
+
+  /**
+   * Preview values corresponding to the current configuration.
+   *
+   */
+  getLiveValues() {
+    // only once
+    if (this.liveData != null){
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.ts:72]: space between `)` and `{`

File: projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.ts [hunk 12]
```diff
@@ -14,19 +14,68 @@
...
+  /** onDestroy hook */
+  ngOnDestroy(): void {
+    this._subscriptions.unsubscribe();
+  }
+
+  /**
+   * Preview values corresponding to the current configuration.
+   *
+   */
+  getLiveValues() {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.ts:70]: `getLiveValues(): void`

**Review by Garfield-fr — APPROVED** [review]

File: projects/admin/src/app/menu/menu-definition/menu-app.ts [hunk 13]
```diff
@@ -215,7 +215,16 @@ export const MENU_APP: IMenuParent[] = [
         access: {
           permissions: [PERMISSIONS.PTTR_ACCESS]
         }
-      }
+      },
+      {
+        name: 'Report Configuration',
+        router_link: ['/', 'records', 'stats_cfg'],
+        attributes: { id: 'stats-cfg-menu' },
+        extras: { iconClass: 'fa fa-cog' },
```
Comment (PascalRepond) [inline projects/admin/src/app/menu/menu-definition/menu-app.ts:223]: `fa-bar-chart` would be nice?

**Review by PascalRepond — APPROVED** [review]

**Review by lauren-d — APPROVED** [review]

File: projects/admin/src/app/menu/menu-definition/menu-app.ts (hunk 13 above)
Comment (jma) [reply projects/admin/src/app/menu/menu-definition/menu-app.ts:223]: `fa-bar-chart` is already used by the main menu.

File: projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.ts [hunk 14]
```diff
@@ -14,19 +14,65 @@
...
+    // only once
+    if (this.liveData != null) {
+      return;
+    }
+    const pid = this.record.metadata.pid;
+    const baseUrl = this._apiService.endpointPrefix;
+    this._http
+      .get(`${baseUrl}/stats_cfg/live/${pid}`)
+      .subscribe((res) => (this.liveData = res));
+  }
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/statistics-cfg-detail-view.component.ts:77]: @Garfield-fr likes to use a dedicated "api.service.ts" for this kind of API call. The principle is "component doesn't need to call backend, just use data to compose a view. Only services must call backend."

File: projects/admin/src/app/routes/statistics-cfg-route.ts [hunk 15]
```diff
@@ -15,65 +15,166 @@
...
+              jsonSchema: JSONSchema7
+            ): FormlyFieldConfig => {
+              const formWidget = jsonSchema.widget;
+              if (
+                formWidget?.formlyConfig?.templateOptions?.fieldMap ===
+                "libraries"
+              ) {
+                field.type = "select";
+                field.hooks = {
+                  ...field.hooks,
```
Comment (zannkukai) [inline projects/admin/src/app/routes/statistics-cfg-route.ts:115]: For readability, maybe i will extract this into a dedicated function. Like `_populateLocationsByCurrentUserLibrary()` function into `items_routes.ts`

File: projects/admin/src/app/routes/statistics-cfg-route.ts [hunk 16]
```diff
@@ -15,65 +15,166 @@
...
+                          return hits.map((hit: any) => {
+                            return {
+                              label: hit.metadata.name,
+                              value: apiService.getRefEndpoint(
+                                "libraries",
+                                hit.metadata.pid
+                              ),
+                            };
+                          });
+                        })
```
Comment (zannkukai) [inline projects/admin/src/app/routes/statistics-cfg-route.ts:158]: cosmetic : first `return` could be removed 
```typescript
map((hits: any[]) => hits.map((hit: any) => {
    return {
      label: hit.metadata.name,
      value: apiService.getRefEndpoint("libraries", hit.metadata.pid)
    };
  });
)
```

**Review by zannkukai — APPROVED** [review]

File: projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/report-data/report-data.component.html [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+-->
+
+<p-table *ngIf="data && data.length > 0; else noResult" [value]="data" [tableStyle]="{ 'min-width': '50rem' }">
+  <ng-template pTemplate="body" let-line>
+    <tr>
+      <td *ngFor="let value of line">{{ value }}</td>
+    </tr>
+  </ng-template>
+</p-table>
```
[the remark this answers is not in the document]
Comment (jma) [reply projects/admin/src/app/record/detail-view/statistics-cfg-detail-view/report-data/report-data.component.html:24]: Yes!

**Pull request merged**